### DISCUSSION TITLE
how to use PIO to replace SPI?

### DISCUSSION THREAD

**Discussion by carl336** [micropython/micropython#14207] [RP2040 / Pico]
Hi, can anybody teach me how to use the PIO function to replace SPI? 
The hardware I am using now is Raspberry Pi Pico with RP2040
 (MicroPython v1.22.0 on 2023-12-27)

Here is the Micropython code I wrote:
https://paste.fastmirror.net/?94d3afc18f71f351#FpA3y3DqZjA5i8kAe9hJ1kCbDLjnv8i1sNfRY7D4MST5

**Comment by rkompass** [comment]
I would start with this:
[https://github.com/raspberrypi/pico-micropython-examples/blob/master/pio/pio_spi.py](https://github.com/raspberrypi/pico-micropython-examples/blob/master/pio/pio_spi.py).

**Comment by rikkenharp** [comment]
If you also are looking for speed you will have to befriend with the DMA too.
I got myself a cheap logic analyzer when I startet programming PIO and DMA, in my case it was a BIG help to see what was really going on.
Maybe the best 8€ I ever have spend :)

**Comment by fdufnews** [comment]
@carl336 this is the second discussion you created with what seems a similar question.
You are talking about SPI but it seems this is not really SPI you are expecting.
Could you be a little more explicit about what you expect. Give the reference of the circuit you want to interface to.
And please copy and paste your code or your links in the message instead of giving link to an external hosting. The links for this type of hosting will be dead sooner or later and this discussion will be completely incomprehensible.
